rfc/ucanto-result.md

```diff
@@ -0,0 +1,210 @@
+## Abstract
+
+Ucanto's `Result` type gives us the ability to use the TypeScript type system to specify the errors a function may return. 
+
+We have started using `Result`s in a variety of different interfaces in our system, but have not been careful to catch and wrap runtime errors with specified error types in all cases. This has lead to functions that require callers to implement two different types of error handling - detection and handling of "error" `Result`s like `{error: {name: 'AnError', message: 'A message describing the error'}}` as well as standard `try...catch` error handling. 
+
+We describe the tradeoffs and suggest that best practice is to coerce all errors into `Result` types in functions that return `Result`.
+
+## `Ucanto.Result`
+
+Ucanto `Result`s are described by the [UCAN invocation specification](https://github.com/ucan-wg/invocation/#6-result) and [implemented as a TypeScript `type` in `@web3-storage/ucanto`](https://github.com/web3-storage/ucanto/blob/main/packages/interface/src/lib.ts#L506). 
+
+Briefly, a `Ucanto.Result` is a TypeScript `Record` with EITHER an `ok` or `error` key. If `ok` is defined, `error` must be `undefined`, and vice-versa. Defining an function that returns `Result` allows us to specify possible error types for that function, which enables conveniences like IDE auto-complete and helps clients understand what types of errors they should expect to handle:
+
+```typescript
+interface PlanGetSuccess {
+  updatedAt: ISO8601Date
+  product: DID
+}
+
+interface PlanNotFound extends Ucanto.Failure {
+  name: 'PlanNotFound'
+}
+
+type PlanGetFailure = PlanNotFound
+
+interface PlansStorage {
+  /**
+   * Get plan information for a customer
+   *
+   * @param account account DID
+   */
+  get: (
+    account: AccountDID
+  ) => Promise<Ucanto.Result<PlanGetSuccess, PlanGetFailure>>
+}
+
+// using a Result from a normal TypeScript function
+async function getProductForAccount(plansStorage: PlansStorage, account: AccountDID): Promise<DID | null> {
+  const result = await plansStorage.get(account)
+  if (result.ok) {
+    return result.ok.product
+  } else if (result.error.name === 'PlanNotFound' ) {
+    return null
+  } else {
+    throw result.error
+  }
```

Comment:
Probably worth mentioning using `Result.try` to de-verboseify this.

Reply:
ah good idea - I'm not sure the result is much less verbose, but it's a good idea to highlight this utility!

here's what I came up with: 

```typescript
async function getProductForAccount(plansStorage: PlansStorage, account: AccountDID): Promise<DID | null> {
  try {
    return Result.try(await plansStorage.get(account)).product
  } catch (err) {
    if (err.name === 'PlanNotFound') {
      return null
    } else {
      throw err
    }
  }
}
```